src/gui/guiFormSpecMenu.cpp

```diff
@@ -4228,153 +4286,384 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
+				// Start picking up items
+				m_selected_item = new GUIInventoryList::ItemSpec(s);
+				m_selected_amount = s_count;
+				m_selected_dragging = true;
 			}
-		break;
+			break;
+		}
+		case BET_OTHER: {
+			// Some other mouse event has occured
+			// Currently only left-double-click should trigger this
+			if (!s.isValid() || event.MouseInput.Event != EMIE_LMOUSE_DOUBLE_CLICK)
+				break;
+
+			// Both the selected item and the hovered item need to be checked
+			// because we don't know exactly when the double-click happened
+			ItemStack slct;
+			if (!m_selected_item && !empty)
+				slct = list_s->getItem(s.i);
+			else if (m_selected_item && (identical || empty))
+				slct = list_selected->getItem(m_selected_item->i);
+
+			// Pickup all of the item from the list
+			if (slct.count > 0) {
+				for (u32 i = 0; i < list_s->getSize(); i++) {
+					// Skip the selected slot
+					if (i == s.i)
+						continue;
+
+					ItemStack item = list_s->getItem(i);
+
+					if (slct.name == item.name && slct.wear == item.wear &&
+							slct.metadata == item.metadata) {
+						// Found a match, check if we can pick it up
+						bool full = false;
+						u16 amount = item.count;
+						ItemStack leftover = slct.addItem(item, m_client->idef());
+						if (!leftover.empty()) {
+							amount -= leftover.count;
+							full = true;
+						}
+
+						if (amount > 0) {
+							IMoveAction *a = new IMoveAction();
+							a->count = amount;
+							a->from_inv = s.inventoryloc;
+							a->from_list = s.listname;
+							a->from_i = i;
+							a->to_inv = s.inventoryloc;
+							a->to_list = s.listname;
+							a->to_i = s.i;
+							m_invmgr->inventoryAction(a);
+
+							if (m_selected_item)
+								m_selected_amount += amount;
+						}
+
+						if (full) // Stack is full, stop
+							break;
+					}
+				}
+			}
+			break;
+		}
 		default:
 			break;
 		}
 
+		// Update left-dragged slots
+		if (m_left_dragging && m_left_drag_stacks.size() > 1) {
+			// The split amount will always at least one, because the number
+			// of slots will never be greater than the selected amount
+			u16 split_amount = m_left_drag_amount / m_left_drag_stacks.size();
+
+			ItemStack stack_from = m_left_drag_stack;
+			m_selected_amount = m_left_drag_amount;
+
+			for (auto &ds : m_left_drag_stacks) {
+				Inventory *inv_to = m_invmgr->getInventory(ds.first->inventoryloc);
+				InventoryList *list_to = inv_to->getList(ds.first->listname);
+
+				if (inv_to == inv_selected && list_to == list_selected &&
+						ds.first->i == m_selected_item->i) {
+					// Adding to the source stack, just change the selected amount
+					m_selected_amount -= split_amount;
+
+				} else {
+					// Reset the stack to it's original state
+					list_to->changeItem(ds.first->i, ds.second);
+
+					// Add the new split to the stack
+					ItemStack add_stack = stack_from;
+					add_stack.count = split_amount;
+					ItemStack leftover = list_to->addItem(ds.first->i, add_stack);
```

Comment:
Q: The list contents apparently change with each iteration. How and where does this trigger the inventory action?

Reply:
Yes, the items are redistributed each time an additional slot is hovered over while dragging. The inventory action only happens when left-click is released. (this is why `inhibit_inventory_revert` is needed to prevent the changes from being reverted)

https://github.com/minetest/minetest/blob/b795b6cdb54a12b0f0101dda064679f1fbfdb6ba/src/gui/guiFormSpecMenu.cpp#L4455-L4493